component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java

```diff
@@ -421,7 +473,22 @@ public String getName() {
      */
     private boolean generateOTPAndSendByEmail(AuthenticationContext context) {
 
-        String username = getUsernameFromContext(context);
+        String username = null;
+        AuthenticatedUser authenticatedUser = null;
+        boolean isLocalUser = TOTPUtil.isLocalUser(context);
+        if (isLocalUser) {
+            username = getUsernameFromContext(context);
+        } else {
+            authenticatedUser = TOTPUtil.getAuthenticatedUser(context);
+            if (authenticatedUser == null) {
+                log.error("Error when generating the totp token due to authenticated federated user " +
+                        "is not found in the context.");
+                return false;
+            }
+            if (isFederatedUser(authenticatedUser)) {
```

Comment:
do we need this check here? doesn't the condition at line #479 check whether user in context is local or federated?

Reply:
Fixed